metricbeat/module/elasticsearch/index/data_xpack.go

```diff
@@ -0,0 +1,332 @@
+		"miss_count":           c.Int("miss_count"),
+	}
+)
+
+func eventsMappingXPack(r mb.ReporterV2, m *MetricSet, info elasticsearch.Info, content []byte) error {
+	err := json.Unmarshal(content, &indicesStruct)
+	if err != nil {
+		return err
+	}
+
+	clusterStateMetrics := []string{"metadata", "routing_table"}
+	clusterState, err := elasticsearch.GetClusterState(m.HTTP, m.HTTP.GetURI(), clusterStateMetrics)
+	if err != nil {
+		return err
+	}
+
+	for name, index := range indicesStruct.Indices {
+		event := mb.Event{}
+		indexStats, err := xpackSchema.Apply(index)
+		if err != nil {
+			continue
+		}
+		indexStats["index"] = name
+
+		err = addClusterStateFields(name, indexStats, clusterState)
+		if err != nil {
+			continue
+		}
+
+		event.RootFields = common.MapStr{
+			"cluster_uuid": info.ClusterID,
+			"timestamp":    common.Time(time.Now()),
+			"interval_ms":  m.Module().Config().Period / time.Millisecond,
+			"type":         "index_stats",
+			"index_stats":  indexStats,
+		}
+
+		event.Index = elastic.MakeXPackMonitoringIndexName(elastic.Elasticsearch)
+		r.Event(event)
+	}
+
+	return nil
+}
+
+func addClusterStateFields(indexName string, indexStats, clusterState common.MapStr) error {
+	indexMetadata, err := getClusterStateMetricForIndex(clusterState, indexName, "metadata")
+	if err != nil {
+		return err
+	}
+
+	indexRoutingTable, err := getClusterStateMetricForIndex(clusterState, indexName, "routing_table")
+	if err != nil {
+		return err
+	}
+
+	// TODO:
+
+	// "index_stats.created",
+	v, err := indexMetadata.GetValue("settings.index.creation_date")
+	if err != nil {
+		return err
+	}
+	c, ok := v.(string)
+	if !ok {
+		return elastic.MakeErrorForMissingField("settings.index.creation_date", elastic.Elasticsearch)
+	}
+
+	created, err := strconv.ParseInt(c, 10, 64)
+	if err != nil {
+		return err
+	}
+	indexStats.Put("created", created)
+
+	// "index_stats.version.created", <--- don't think this is being used in the UI, so can we skip it?
+	// "index_stats.version.upgraded", <--- don't think this is being used in the UI, so can we skip it?
+
+	// "index_stats.status",
+	s, err := indexRoutingTable.GetValue("shards")
+	if err != nil {
+		return err
+	}
+	shards, ok := s.(map[string]interface{})
+	if !ok {
+		return elastic.MakeErrorForMissingField("shards", elastic.Elasticsearch)
+	}
+
+	status, err := getIndexStatus(shards)
+	if err != nil {
+		return err
+	}
+	indexStats.Put("status", status)
+
+	// "index_stats.shards.total",
+	// "index_stats.shards.primaries",
+	// "index_stats.shards.replicas",
+	// "index_stats.shards.active_total",
+	// "index_stats.shards.active_primaries",
+	// "index_stats.shards.active_replicas",
+	// "index_stats.shards.unassigned_total",
+	// "index_stats.shards.unassigned_primaries",
+	// "index_stats.shards.unassigned_replicas",
+	// "index_stats.shards.initializing",
+	// "index_stats.shards.relocating",
+	shardStats, err := getIndexShardStats(shards)
+	if err != nil {
+		return err
```

Comment:
This function returns errors in several places, you may want to make use of [`errors.Wrap()`](https://godoc.org/github.com/pkg/errors#hdr-Adding_context_to_an_error) or similar to give more info about when it failed, for example here it could be `return errors.Wrap(err, "failed to get index shard stats")`.